File: src/lexer/lexer.zig
const std = @import("std");

pub const Token = struct {
    kind: Kind,
    span: Span,

    const Self = @This();

    pub const Span = struct {
        start: usize,
        end: usize,
    };

    pub const Kind = enum {
        None, 

        Identifier,
        String,
        Integer,
        Float,
        Operator,

        Equals,
        ColonEquals,
        SemiColon,
        Colon,
        ColonColon,
        FatArrow,
        Comma,
        Dot,
        LeftParen,
        RightParen,
        LeftBrace,
        RightBrace,
        LeftBracket,
        RightBracket,

        TypeKeyword,
        ClassKeyword,
        InstanceKeyword,
        CaseKeyword,
        LetKeyword,
        InKeyword,
        IfKeyword,
        ThenKeyword,
        ElseKeyword,
        ElseIfKeyword,
        DefKeyword,
        ConstKeyword,
        WhereKeyword,
        DoKeyword,
    };

    pub const keywords = std.StaticStringMap(Kind).initComptime(.{
        .{ "type", .TypeKeyword },
        .{ "class", .ClassKeyword },
        .{ "instance", .InstanceKeyword },
        .{ "case", .CaseKeyword },
        .{ "let", .LetKeyword },
        .{ "in", .InKeyword },
        .{ "if", .IfKeyword },
        .{ "then", .ThenKeyword },
        .{ "else", .ElseKeyword },
        .{ "elseif", .ElseIfKeyword },
        .{ "def", .DefKeyword },
        .{ "const", .ConstKeyword },
        .{ "where", .WhereKeyword },
        .{ "do", .DoKeyword },
    });

    pub fn get_keyword(buff: []const u8) ?Kind {
        return keywords.get(buff);
    }

    pub fn get_lexeme(self: Self) ?[]const u8 {
        return switch (self.kind) {
            // these all have a payload
            .None, .Identifier, .Integer, .Float, .String, .Boolean, .Operator => null,

            // these are constant.
            .Equals => "=",
            .ColonEquals => ":=",
            .SemiColon => ";",
            .Colon => ":",
            .ColonColon => "::",
            .FatArrow => "=>",
            .Comma => ",",
            .Dot => ".",
            .LeftParen => "(",
            .RightParen => ")",
            .LeftBrace => "{",
            .RightBrace => "}",
            .LeftBracket => "[",
            .RightBracket => "]",

            .TypeKeyword => "type",
            .ClassKeyword => "class",
            .InstanceKeyword => "instance",
            .CaseKeyword => "case",
            .LetKeyword => "let",
            .InKeyword => "in",
            .IfKeyword => "if",
            .ThenKeyword => "then",
            .ElseKeyword => "else",
            .ElseIfKeyword => "elseif",
            .DefKeyword => "def",
            .ConstKeyword => "const",
            .WhereKeyword => "where",
            .DoKeyword => "do",
        };
    }
};

pub const Lexer = struct {
    const Self = @This();

    src: [:0]const u8,
    pos: usize,
    ctok: ?Token,

    pub fn init(src: [:0]const u8) Lexer {
        return Lexer{
            .src = src,
            .index = 0,
        };
    }

    pub const State = enum {
        Start,

        Identifier,
        Integer,
        Float,

        Comment,
    };

    pub fn next(self: *Self) Token {
        if (self.ctok) |token| {
            self.ctok = null;
            return token;
        }

        var result = Token{
            .kind = .None,
            .span = .{
                .start = self.pos,
                .end = undefined,
            },
        };

        while (true) : (self.pos += 1) {
        }
    }
};

File: test/lexer.ml
open OUnit2

let test string tokens =
  let open Jace in
  let code = Code.init string in
  let lexer = Lexer.init code in
  let rec test' lexer tokens =
    match tokens with
      | (kind, str) :: tail ->
        let lexer, token = Lexer.next lexer in
        assert_equal kind token.kind;
        assert_equal str (Code.read code token.span);
        test' lexer tail
      | _ -> ()
  in
  test' lexer tokens
;;

let test_keywords _ =
  let open Jace.Token in
  test "type class instance def where let in case if then else elseif ref inherit"
    [ (TypeKeyword, "type")
    ; (ClassKeyword, "class")
    ; (InstanceKeyword, "instance")
    ; (DefKeyword, "def")
    ; (WhereKeyword, "where")
    ; (LetKeyword, "let")
    ; (InKeyword, "in")
    ; (CaseKeyword, "case")
    ; (IfKeyword, "if")
    ; (ThenKeyword, "then")
    ; (ElseKeyword, "else")
    ; (ElseIfKeyword, "elseif")
    ; (RefKeyword, "ref")
    ; (InheritKeyword, "inherit")
    ; (Eof, "")
    ]
;;

let test_identifiers _ =
  let open Jace.Token in
  test "foo _bar baz123 foobarbaz_123 __wow"
    [ (Identifier, "foo")
    ; (Identifier, "_bar")
    ; (Identifier, "baz123")
    ; (Identifier, "foobarbaz_123")
    ; (Identifier, "__wow")
    ]
;;

(* TODO handle invalid numbers and test them *)
let test_numbers _ =
  let open Jace.Token in
  test "123 1.23 0.23 10000 0.00004 0"
    [ (Integer, "123")
    ; (Float, "1.23")
    ; (Float, "0.23")
    ; (Integer, "10000")
    ; (Float, "0.00004")
    ; (Integer, "0")
    ; (Eof, "")
    ]
;;

let test_strings _ =
  let open Jace.Token in
  test {|"Hello World!" "Escaped: \"Foo\"" "Some more escapes:\t\n\r"|}
    [ (String, {|"Hello World!"|})
    ; (String, {|"Escaped: \"Foo\""|})
    ; (String, {|"Some more escapes:\t\n\r"|})
    ; (Eof, "")
    ]
;;

let test_invalid_strings _ =
  let open Jace.Token in
  (* bad strings still match, but are still considered invalid. *)
  test {|"Bad string then next line:
         "Bad string then eof
         "Invalid escape sequence: \a <- this guy"|}
    [ (InvalidString, {|"Bad string then next line:|})
    ; (InvalidString, {|"Bad string then eof|})
    ; (InvalidString, {|"Invalid escape sequence: \a <- this guy"|})
    ; (Eof, "")
    ]
;;

let test_delimeters _ =
  let open Jace.Token in
  test "; , \\ => ( ) { } [ ]"
    [ (SemiColon, ";")
    ; (Comma, ",")
    ; (BackSlash, "\\")
    ; (FatArrow, "=>")
    ; (LeftParen, "(")
    ; (RightParen, ")")
    ; (LeftBrace, "{")
    ; (RightBrace, "}")
    ; (LeftBracket, "[")
    ; (RightBracket, "]")
    ; (Eof, "")
    ]
;;

let test_delimeters_1 _ =
  let open Jace.Token in
  test "(())=>,\\;[]{}(=>)"
    [ (LeftParen, "(")
    ; (LeftParen, "(")
    ; (RightParen, ")")
    ; (RightParen, ")")
    ; (FatArrow, "=>")
    ; (Comma, ",")
    ; (BackSlash, "\\")
    ; (SemiColon, ";")
    ; (LeftBracket, "[")
    ; (RightBracket, "]")
    ; (LeftBrace, "{")
    ; (RightBrace, "}")
    ; (LeftParen, "(")
    ; (FatArrow, "=>")
    ; (RightParen, ")")
    ; (Eof, "")
    ]
;;

let test_reserved_operators _ =
  let open Jace.Token in
  test "= := ::"
    [ (Equal, "=")
    ; (ColonEqual, ":=")
    ; (ColonColon, "::")
    ; (Eof, "")
    ]
;;

let test_default_implemented_operators _ =
  let open Jace.Token in
  test "+ - * / : > >= < <= ^ == != . .."
    [ (Operator, "+")
    ; (Operator, "-")
    ; (Operator, "*")
    ; (Operator, "/")
    ; (Operator, ":")
    ; (Operator, ">")
    ; (Operator, ">=")
    ; (Operator, "<")
    ; (Operator, "<=")
    ; (Operator, "^")
    ; (Operator, "==")
    ; (Operator, "!=")
    ; (Operator, ".")
    ; (Operator, "..")
    ; (Eof, "")
    ]
;;

let test_primitive_operators _ =
  let open Jace.Token in
  test {| @and @not @or @intAdd @intSub @intMul @intDiv
          @intExp @floatAdd @floatSub @floatMul @floatDiv
          @floatExp @strLen @strAppend @strConcat @strCharAt
          @listLen @listAppend @listIndex @recordIndex @recordLen 
          @ref @deref|}
    [ (AndOp, "@and")    
    ; (NotOp, "@not")    
    ; (OrOp, "@or")      
    ; (IntAddOp, "@intAdd") 
    ; (IntSubOp, "@intSub") 
    ; (IntMulOp, "@intMul") 
    ; (IntDivOp, "@intDiv") 
    ; (IntExpOp, "@intExp") 
    ; (FloatAddOp, "@floatAdd") 
    ; (FloatSubOp, "@floatSub") 
    ; (FloatMulOp, "@floatMul") 
    ; (FloatDivOp, "@floatDiv") 
    ; (FloatExpOp, "@floatExp") 
    ; (StrLenOp, "@strLen")       
    ; (StrAppendOp, "@strAppend") 
    ; (StrConcatOp, "@strConcat") 
    ; (StrCharAtOp, "@strCharAt")   
    ; (ListLenOp, "@listLen")        
    ; (ListAppendOp, "@listAppend")  
    ; (ListIndexOp, "@listIndex")    
    ; (RecordIndexOp, "@recordIndex")
    ; (RecordLenOp, "@recordLen")
    ; (RefOp, "@ref")
    ; (DerefOp, "@deref")
    ; (Eof, "")
    ]
;;

let test_invalid_primitive_operators _ =
  let open Jace.Token in
  test "@foo @bar @baz"
    [ (InvalidPrimitiveOp, "@foo")
    ; (InvalidPrimitiveOp, "@bar")
    ; (InvalidPrimitiveOp, "@baz")
    ; (Eof, "")
    ]
;;

let tests = "Lexer tests"
  >:::
    [ "keywords" >:: test_keywords 
    ; "numbers" >:: test_numbers
    ; "delimeters" >:: test_delimeters 
    ; "delimeters_1" >:: test_delimeters_1 
    ; "reserved_operators" >:: test_reserved_operators
    ; "default_operators" >:: test_default_implemented_operators
    ; "primitive_operators" >:: test_primitive_operators
    ; "invalid_primitive_operators" >:: test_invalid_primitive_operators
    ; "identifiers" >:: test_identifiers
    ; "strings" >:: test_strings
    ; "invalid_strings" >:: test_invalid_strings
    ]
;;

let () = run_test_tt_main tests
